Bot.Database/TownDatabase.cs

```diff
@@ -38,35 +38,61 @@ private static MongoTownRecord RecordFromTown(ITown town, IMember author)
 				StorytellerRoleId = town.StorytellerRole?.Id ?? 0,
 				VillagerRole = town.VillagerRole?.Name,
 				VillagerRoleId = town.VillagerRole?.Id ?? 0,
-				AuthorName = author.DisplayName,
-				Author = author.Id,
+				AuthorName = authorName,
+				Author = authorId,
 				Timestamp = DateTime.Now,
 			};
-
 		}
 
-        public async Task<bool> AddTown(ITown town, IMember author)
+		private static MongoTownRecord RecordFromTown(ITown town, IMember author) => RecordFromTownAndAuthorInfo(town, author.Id, author.DisplayName);
+
+        public async Task<bool> AddTownAsync(ITown town, IMember author)
         {
 			var newRec = RecordFromTown(town, author);
 
 			// TODO: error check this record?
-			await m_collection.InsertOneAsync(newRec);
-
+			await UpdateRecordAsync(newRec, false);
 			return true;
         }
 
-        public async Task<ITownRecord?> GetTownRecord(ulong guildId, ulong channelId)
-		{
-			// Build a filter for the specific document we want
-			var builder = Builders<MongoTownRecord>.Filter;
-			var filter = builder.Eq(x => x.GuildId, guildId) & builder.Eq(x => x.ControlChannelId, channelId);
+		public async Task<bool> UpdateTownAsync(ITown town)
+        {
+			if (town.Guild == null || town.ControlChannel == null)
+				return false;
+
+            if (await GetTownRecordAsync(town.Guild.Id, town.ControlChannel.Id) is not MongoTownRecord oldRec)
+                return false;
 
-			// Get the first match
-			var document = await m_collection.Find(filter).FirstOrDefaultAsync();
-			return document;
+            var newRec = RecordFromTownAndAuthorInfo(town, oldRec.Author, oldRec.AuthorName);
+			newRec._id = oldRec._id;
+
+			await UpdateRecordAsync(newRec, true);
+			return true;
 		}
 
-		public async Task<IEnumerable<ITownRecord>> GetTownRecords(ulong guildId)
+		private Task UpdateRecordAsync(MongoTownRecord record, bool upsert)
+        {
+			return m_collection.ReplaceOneAsync(GetTownMatchFilter(record.GuildId, record.ControlChannelId), record, new ReplaceOptions() { IsUpsert = upsert });
```

Comment:
There seems to be a problem here.

If you try to add a town that already exists, it just won't add here because I put Upsert=false
But if I set Upset=true, it will throw an exception because the immutable _id field will be changed.
Any ideas how to fix this?

On line 67 I "fixed" this for the Update case by setting the _id of the new record = the _id of the old record, but this seems goofy. It seems to me that if I can find the Town using the Filter, I shouldn't have to bother with the ID.

How can we make it so the ID doesn't get changed when a ReplaceOneAsync call happens?

Reply:
We shouldn't have _id in our serialized objects at all... we should let Mongo handle that on its side